Remove componentWill* methods

hawkrives opened the pull request:
Closes #2430

React is removing the `componentWill*` methods: https://github.com/reactjs/rfcs/blob/master/text/0006-static-lifecycle-methods.md

The impetus for this is a future React release which will switch to "async rendering", wherein they will do cool things like _interrupt the rendering of a component to render a higher-priority thing_. But, to do that, they can't guarantee that `componentWill*` will only be called once, so they're being deprecated.

Here's your migration chart:

```js
  UNSAFE_componentWillMount() {
    // New name for componentWillMount()
    // Indicates that this method can be unsafe for async rendering.
    // Prefer componentDidMount() instead.
  }

  UNSAFE_componentWillUpdate(nextProps, nextState) {
    // New name for componentWillUpdate()
    // Indicates that this method can be unsafe for async rendering.
    // Prefer componentDidUpdate() instead.
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    // New name for componentWillReceiveProps()
    // Indicates that this method can be unsafe for async rendering.
    // Prefer static getDerivedStateFromProps() instead.
  }
```

The trickiest one is `getDerivedStateFromProps`; here's an example migration I did:

```diff
-	componentWillReceiveProps(nextProps: Props) {
-		this.setState(state => ({
+	static getDerivedStateFromProps(nextProps: Props, prevState: State) {
+		return {
			buildings: groupBuildings(
-				state.allBuildings,
+				prevState.allBuildings,
				nextProps.favoriteBuildings,
			),
-		}))
+		}
	}
```

For some of the other `componentWillReceiveProps`-esque methods we have, like `_performSearch` in CourseSearch, I extracted the bulk of the logic into a function that lives outside of the class entirely and returns an object instead of calling `this.setState` directly. Then I call that function both from `static deriveStateFromProps` and from `this._performSearch`.

For example: 

https://github.com/StoDevX/AAO-React-Native/blob/9269c0149e8603132cd15b7e6ab8ac2a40eb55bd/source/views/sis/course-search/search.js#L205-L214

https://github.com/StoDevX/AAO-React-Native/blob/9269c0149e8603132cd15b7e6ab8ac2a40eb55bd/source/views/sis/course-search/search.js#L128-L138

I know there are contradictory labels on this PR, but they're accurate – it's ready for review, but it needs to actually be run on a sim before approving. I haven't yet reinstalled Xcode on my home machine, so I only think that this works.

hawkrives committed:
s/componentWillMount/componentDidMount/

hawkrives committed:
add two new lint rules to avoid setState in didUpdate/Mount

hawkrives committed:
avoid storing the setInterval id in state

just attach it to the class, like a component ref. It has no bearing on
what React does, so it doesn't make sense to put it in state, I think.

hawkrives committed:
make eslint happy by using a callback instead of an async function

because technically that's not supposed to be async

hawkrives committed:
remove duplicate componentDidMount from Datepicker

hawkrives committed:
update DatePicker to use getDerivedStateFromProps

hawkrives committed:
sort rules alphabetically

hawkrives committed:
add no-access-state-in-setstate rule

hawkrives committed:
fix??? bonapp menu picker component???

hawkrives committed:
port StatefulList to getDerivedStateFromProps

hawkrives committed:
port HoursRow to getDerivedStateFromProps

a bit more information: I extracted the derivation process into a
standalone function that both the initial this.state setup and the
getDerivedStateFromProps function call, thus actually removing 20 lines
of code from this file.

hawkrives committed:
replace componentWillReceiveProps in MenuBonapp

this one is a bit different: I switched it to componentDidUpdate, and
also made it cache the current menu id in state.

hawkrives committed:
prettify hoursrow

hawkrives committed:
update BusLine to avoid componentWillReceiveProps

this follows the same pattern as BuildingHoursRow

hawkrives committed:
move EditHomeRow away from componentWillReceiveProps to componentDidUpdate

also extracts the Animated values from state, since they're not managed
by setState

hawkrives committed:
tweak course-search/search.js formatting

hawkrives committed:
rename onFocus/onCancel to handle*

hawkrives committed:
migrate course-search/search to getDerivedStateFromProps

following the BuildingHoursRow plan

hawkrives committed:
port fancy-menu to getDerivedStateFromProps

hawkrives committed:
port CourseSearch/list to getDerivedStateFromProps

hawkrives committed:
note that _webview may be null on Player

hawkrives committed:
proxy messages through local state in StreamPlayer

drewvolz reviewed (approved):
Builds and runs. :shipit: 👍

drewvolz reviewed (dismissed):
I realize now that some things need tweaking

hawkrives commented on source/views/menus/components/fancy-menu.js:
```diff
@@ -66,26 +66,21 @@ class FancyMenu extends React.PureComponent<Props> {
+	static getDerivedStateFromProps(nextProps: Props) {
+		// prevent ourselves from overwriting the filters from redux on mount
+		if (nextProps.filters.length) {
 			return
```
This needs to be "return null"

hawkrives commented on source/views/sis/course-search/search.js:
```diff
@@ -76,7 +125,20 @@ class CourseSearchView extends React.PureComponent<Props, State> {
+				allCourses: nextProps.allCourses,
+				updateRecentSearches: nextProps.updateRecentSearches,
+			})
+		}
```
Add "return null" after this line, so the method returns null if the condition doesn't pass.

hawkrives committed:
attempt to fix use of state in components with no state

hawkrives committed:
fix bugsnag warning about sending > 30char breadcrumb names

hawkrives committed:
fix moment deprecation warning (moment#offset -> moment#utcOffset)

hawkrives committed:
properly return null from Datepicker getDerivedStateFromProps

hawkrives committed:
debounce reordering events from the edithome listview

hawkrives committed:
remove state from editHomeRow

hawkrives committed:
Merge branch 'master' into remove-cwm

* master: (52 commits)
  chore(package): update lockfile
  fix(package): update react-native-google-analytics-bridge to version 5.6.3
  Remove extra line at top
  Spring break notice
  update Gemfile.lock
  Bundle Update on 2018-03-21
  chore(package): update lockfile
  fix(package): update react-native-google-analytics-bridge to version 5.6.2
  chore(package): update lockfile
  chore(package): update danger to version 3.3.1
  chore(package): update lockfile
  chore(package): update eslint to version 4.19.1
  chore(package): update lockfile
  chore(package): update ajv to version 6.4.0
  chore(package): update lockfile
  fix(package): update react-native-search-bar to version 3.4.1
  chore(package): update lockfile
  fix(package): update react-native to version 0.54.3
  chore(package): update lockfile
  chore(package): update directory-tree to version 2.1.0
  ...

hawkrives committed:
remove need for maintaining current player state in Player

resolves the last remaining bug in this branch, where you couldn't play
the stream.

hawkrives committed:
remove State parameter from Player
